lib/main.rb

```diff
@@ -0,0 +1,24 @@
+ALPHABET = ('a'..'z').to_a.freeze
+
+def change_letter(letter, shift, memory)
+  return memory[letter] if memory.key?(letter)
+
+  current_index = ALPHABET.find_index(letter)
+  new_index = (current_index + shift) % 26
+  new_letter = ALPHABET[new_index]
+  memory[letter] = new_letter
+
+  new_letter
+end
+
+def caesar_cipher(string, shift)
+  memory = {}
+
+  string.split('').map do |char|
+    if ALPHABET.include?(char)
+      change_letter(char, shift, memory)
+    else
+      char
+    end
+  end.join
+end
```

Comment:
Just a small refactor! To avoid passing the `memory` hash as argument to `change_letter` and mutate the hash inside because it's passed by reference, I think it would be better to operate the hash in this same method
```suggestion 
def caesar_cipher(string, shift)
  memory = {}

  string.split('').map do |char|
    result = memory[letter]
    result ||= ALPHABET.include?(char) ? change_letter(char, shift) : char
    memory[letter] = result
    
    result
  end.join
end
```

Reply:
I love it!